library/network/src/modules/PortAliases.rb

```diff
@@ -141,175 +65,114 @@ def AllowedPortNameOrNumber
+    # Returns the port for requested service (if any)
+    #
+    # @note when given argument looks like a digit, it will be returned after a proper conversion
+    #
+    # @param needle [String] the name or alias of the service
+    # @return [Integer, nil] a port number if any
+    def GetPortNumber(needle)
+      return needle.to_i if numeric?(needle)
 
-      alias_found = port.to_i
+      service = find_by_alias(needle)
+      service&.port
+    end
 
-      # store results for later requests
-      Ops.set(@SERVICE_NAME_TO_PORT, port_name, alias_found)
+    publish function: :IsAllowedPortName, type: "boolean (string)"
+    publish function: :AllowedPortNameOrNumber, type: "string ()"
+    publish function: :GetListOfServiceAliases, type: "list <string> (string)"
+    publish function: :IsKnownPortName, type: "boolean (string)"
+    publish function: :GetPortNumber, type: "integer (string)"
 
-      alias_found
-    rescue Errno, IOError => e
-      log.error "failed to read /etc/services #{e.inspect}"
-    end
+  private
 
-    # Function returns list of aliases (port-names and port-numbers) for
-    # requested port-number or port-name. Also the requested name or port is returned.
+    # Returns the collection of services
     #
-    # @param [String] port-number or port-name
-    # @return  [Array] [string] of aliases
-    def GetListOfServiceAliases(port)
-      service_aliases = [port]
-      port_number = nil
-
-      # service is a port number
-      if Builtins.regexpmatch(port, "^[0123456789]+$")
-        port_number = Builtins.tointeger(port)
+    # Results are cached, so repeated requests are answered faster.
+    #
+    # @see #services_database
+    #
+    # @return [Hash<Integer => Service>] a list of services indexed by the port
+    def services
+      @services ||= services_database.each_with_object({}) do |service, result|
+        # Extract values splitting the line by spaces or `/`
+        name, port, _protocol, aliases = service.split(/\s+|\//)
 
-        sport_name = Ops.get(@SERVICE_PORT_TO_NAME, port_number) do
-          LoadAndReturnPortToName(port_number)
-        end
+        port = port.to_i
+        aliases = [name, aliases&.split].flatten.compact
 
-        if !sport_name.nil?
-          service_aliases = Convert.convert(
-            Builtins.union(
-              service_aliases,
-              Builtins.splitstring(sport_name, " ")
-            ),
-            from: "list",
-            to:   "list <string>"
-          )
-        end
-        # service is a port name, any space isn't allowed
-      elsif IsAllowedPortName(port)
-        found_alias_port = Ops.get(@SERVICE_NAME_TO_PORT, port) do
-          LoadAndReturnNameToPort(port)
-        end
-        if !found_alias_port.nil?
-          service_aliases = Builtins.add(
-            service_aliases,
-            Builtins.tostring(found_alias_port)
-          )
+        result[port] ||= Service.new(port, [])
+        result[port].aliases |= aliases
```

Comment:
Why not `||=` (for consistency)?

Reply:
Because its an union. That code add the new aliases to the previous one.

Maybe using the _long_ syntax could help to avoid the confusion.

```ruby
result[port].aliases = result[port].aliases.union(aliases)
```

Although the naming is a bit confusing too.

Reply:
Ah, ok, I got confused. Then it is OK.